### ISSUE TITLE
detach the current TUI (nvim continues headlessly)

### ISSUE THREAD

**Issue by eloytoro** [neovim/neovim#23093]
### Problem

When connected to a remote ui executing `:q` will close the remote server, I guess this makes sense in the way that the command is being run by the backend and not the UI, but it's not intuitive or desired, I wish to quit without killing the backend

### Steps to reproduce

nvim --listen 127.0.0.1:8888 --headless
nvim --server 127.0.0.1:8888 --remote-ui
:q

### Expected behavior

The UI should disconnect from the server but not to kill it

### Neovim version (nvim -v)

0.10.x

### Vim (not Nvim) behaves the same?

no

### Operating system/version

macOS 13.2.1

### Terminal name/version

iTerm

### $TERM environment variable

xterm-256color

### Installation

homebrew

**Comment by zeertzjq** [comment]
I don't think this should be changed.

**Comment by erw7** [comment]
> I don't think this should be changed.

I don't think we should try to detach the `tui` with `:q`, but a way to detach the current `tui` should be implemented. 

Even now, it is possible to terminate only `tui` by using `call chanclose(id)` with the channel id obtained by `echo nvim_list_uis()` as an argument. However, it is not easy to use because users need to remember the correspondence between `tui` and channel `id`.

To begin with, #18375 has the following disadvantages, 

- Increased memory usage
- Startup time is increased by the amount of child processes that need to be forked
- Lots of bugs ;-<

so it is unlikely to be beneficial to users unless it is used as follows

```
nohup nvim --listen /tmp/nvim.pipe --headless > /dev/null 2>&1 0< /dev/null &!
nvim --server /tmp/nvim.pipe --remote-ui
```

In that case, we would love to have this feature.

**Comment by mikesmithgh** [comment]
> > I don't think this should be changed.
> 
> I don't think we should try to detach the `tui` with `:q`, but a way to detach the current `tui` should be implemented.
> 
> Even now, it is possible to terminate only `tui` by using `call chanclose(id)` with the channel id obtained by `echo nvim_list_uis()` as an argument. However, it is not easy to use because users need to remember the correspondence between `tui` and channel `id`.
> 
> To begin with, #18375 has the following disadvantages,
> 
> * Increased memory usage
> * Startup time is increased by the amount of child processes that need to be forked
> * Lots of bugs ;-<
> 
> so it is unlikely to be beneficial to users unless it is used as follows
> 
> ```
> nohup nvim --listen /tmp/nvim.pipe --headless > /dev/null 2>&1 0< /dev/null &!
> nvim --server /tmp/nvim.pipe --remote-ui
> ```
> 
> In that case, we would love to have this feature.

Thanks for the info. I came up with a function to disconnect all remote-ui sessions:

```lua
  vim.keymap.set('n', '<leader>q', function()
    for _, ui in pairs(vim.api.nvim_list_uis()) do
      if ui.chan and not ui.stdout_tty then
        vim.fn.chanclose(ui.chan)
      end
    end
  end, { noremap = true })
```

This is working for my use case, but I am not 100% sure that ui.stdout_tty is the correct condition to check. But, if I use your nohup commands in combination with this I can have a long living neovim session. Hopefully this helps someone 🙂

**Comment by cosminadrianpopescu** [comment]
> Even now, it is possible to terminate only tui by using call chanclose(id) with the channel id obtained by echo nvim_list_uis() as an argument. However, it is not easy to use because users need to remember the correspondence between tui and channel id.

Is there a way, when connecting from 2 terminals to the same nvim headless server to know which channel belong to which terminal. The scenario would be when I try to disconnect from a single terminal to not have to disconnect the other also. 

The way I understand it right now, I can do `call chanclose(id)`. But if the `nvim_list_uis()` returns more than one ui, how would I know which channel to close?

**Comment by justinmk** [comment]
> if the `nvim_list_uis()` returns more than one ui, how would I know which channel to close?

`:help nvim_set_client_info()`

**Comment by borring** [comment]
> > if the `nvim_list_uis()` returns more than one ui, how would I know which channel to close?
> 
> `:help nvim_set_client_info()`

I see this function in the help docs, but I don't know how to call it. It doesn't seem to be vimscript, and it's not available in `vim.api`

Currently on `v0.9.2` and my available commands are listed:

<details><summary>Click to Expand</summary>

<p>

```
{
  nvim__buf_redraw_range = <function 1>,
  nvim__buf_stats = <function 2>,
  nvim__get_lib_dir = <function 3>,
  nvim__get_runtime = <function 4>,
  nvim__id = <function 5>,
  nvim__id_array = <function 6>,
  nvim__id_dictionary = <function 7>,
  nvim__id_float = <function 8>,
  nvim__inspect_cell = <function 9>,
  nvim__runtime_inspect = <function 10>,
  nvim__screenshot = <function 11>,
  nvim__stats = <function 12>,
  nvim__unpack = <function 13>,
  nvim_buf_add_highlight = <function 14>,
  nvim_buf_attach = <function 15>,
  nvim_buf_call = <function 16>,
  nvim_buf_clear_highlight = <function 17>,
  nvim_buf_clear_namespace = <function 18>,
  nvim_buf_create_user_command = <function 19>,
  nvim_buf_del_extmark = <function 20>,
  nvim_buf_del_keymap = <function 21>,
  nvim_buf_del_mark = <function 22>,
  nvim_buf_del_user_command = <function 23>,
  nvim_buf_del_var = <function 24>,
  nvim_buf_delete = <function 25>,
  nvim_buf_get_changedtick = <function 26>,
  nvim_buf_get_commands = <function 27>,
  nvim_buf_get_extmark_by_id = <function 28>,
  nvim_buf_get_extmarks = <function 29>,
  nvim_buf_get_keymap = <function 30>,
  nvim_buf_get_lines = <function 31>,
  nvim_buf_get_mark = <function 32>,
  nvim_buf_get_name = <function 33>,
  nvim_buf_get_number = <function 34>,
  nvim_buf_get_offset = <function 35>,
  nvim_buf_get_option = <function 36>,
  nvim_buf_get_text = <function 37>,
  nvim_buf_get_var = <function 38>,
  nvim_buf_is_loaded = <function 39>,
  nvim_buf_is_valid = <function 40>,
  nvim_buf_line_count = <function 41>,
  nvim_buf_set_extmark = <function 42>,
  nvim_buf_set_keymap = <function 43>,
  nvim_buf_set_lines = <function 44>,
  nvim_buf_set_mark = <function 45>,
  nvim_buf_set_name = <function 46>,
  nvim_buf_set_option = <function 47>,
  nvim_buf_set_text = <function 48>,
  nvim_buf_set_var = <function 49>,
  nvim_buf_set_virtual_text = <function 50>,
  nvim_call_dict_function = <function 51>,
  nvim_call_function = <function 52>,
  nvim_chan_send = <function 53>,
  nvim_clear_autocmds = <function 54>,
  nvim_cmd = <function 55>,
  nvim_command = <function 56>,
  nvim_command_output = <function 57>,
  nvim_create_augroup = <function 58>,
  nvim_create_autocmd = <function 59>,
  nvim_create_buf = <function 60>,
  nvim_create_namespace = <function 61>,
  nvim_create_user_command = <function 62>,
  nvim_del_augroup_by_id = <function 63>,
  nvim_del_augroup_by_name = <function 64>,
  nvim_del_autocmd = <function 65>,
  nvim_del_current_line = <function 66>,
  nvim_del_keymap = <function 67>,
  nvim_del_mark = <function 68>,
  nvim_del_user_command = <function 69>,
  nvim_del_var = <function 70>,
  nvim_echo = <function 71>,
  nvim_err_write = <function 72>,
  nvim_err_writeln = <function 73>,
  nvim_eval = <function 74>,
  nvim_eval_statusline = <function 75>,
  nvim_exec = <function 76>,
  nvim_exec2 = <function 77>,
  nvim_exec_autocmds = <function 78>,
  nvim_feedkeys = <function 79>,
  nvim_get_all_options_info = <function 80>,
  nvim_get_autocmds = <function 81>,
  nvim_get_chan_info = <function 82>,
  nvim_get_color_by_name = <function 83>,
  nvim_get_color_map = <function 84>,
  nvim_get_commands = <function 85>,
  nvim_get_context = <function 86>,
  nvim_get_current_buf = <function 87>,
  nvim_get_current_line = <function 88>,
  nvim_get_current_tabpage = <function 89>,
  nvim_get_current_win = <function 90>,
  nvim_get_hl = <function 91>,
  nvim_get_hl_by_id = <function 92>,
  nvim_get_hl_by_name = <function 93>,
  nvim_get_hl_id_by_name = <function 94>,
  nvim_get_keymap = <function 95>,
  nvim_get_mark = <function 96>,
  nvim_get_mode = <function 97>,
  nvim_get_namespaces = <function 98>,
  nvim_get_option = <function 99>,
  nvim_get_option_info = <function 100>,
  nvim_get_option_info2 = <function 101>,
  nvim_get_option_value = <function 102>,
  nvim_get_proc = <function 103>,
  nvim_get_proc_children = <function 104>,
  nvim_get_runtime_file = <function 105>,
  nvim_get_var = <function 106>,
  nvim_get_vvar = <function 107>,
  nvim_input = <function 108>,
  nvim_input_mouse = <function 109>,
  nvim_list_bufs = <function 110>,
  nvim_list_chans = <function 111>,
  nvim_list_runtime_paths = <function 112>,
  nvim_list_tabpages = <function 113>,
  nvim_list_uis = <function 114>,
  nvim_list_wins = <function 115>,
  nvim_load_context = <function 116>,
  nvim_notify = <function 117>,
  nvim_open_term = <function 118>,
  nvim_open_win = <function 119>,
  nvim_out_write = <function 120>,
  nvim_parse_cmd = <function 121>,
  nvim_parse_expression = <function 122>,
  nvim_paste = <function 123>,
  nvim_put = <function 124>,
  nvim_replace_termcodes = <function 125>,
  nvim_select_popupmenu_item = <function 126>,
  nvim_set_current_buf = <function 127>,
  nvim_set_current_dir = <function 128>,
  nvim_set_current_line = <function 129>,
  nvim_set_current_tabpage = <function 130>,
  nvim_set_current_win = <function 131>,
  nvim_set_decoration_provider = <function 132>,
  nvim_set_hl = <function 133>,
  nvim_set_hl_ns = <function 134>,
  nvim_set_hl_ns_fast = <function 135>,
  nvim_set_keymap = <function 136>,
  nvim_set_option = <function 137>,
  nvim_set_option_value = <function 138>,
  nvim_set_var = <function 139>,
  nvim_set_vvar = <function 140>,
  nvim_strwidth = <function 141>,
  nvim_tabpage_del_var = <function 142>,
  nvim_tabpage_get_number = <function 143>,
  nvim_tabpage_get_var = <function 144>,
  nvim_tabpage_get_win = <function 145>,
  nvim_tabpage_is_valid = <function 146>,
  nvim_tabpage_list_wins = <function 147>,
  nvim_tabpage_set_var = <function 148>,
  nvim_win_call = <function 149>,
  nvim_win_close = <function 150>,
  nvim_win_del_var = <function 151>,
  nvim_win_get_buf = <function 152>,
  nvim_win_get_config = <function 153>,
  nvim_win_get_cursor = <function 154>,
  nvim_win_get_height = <function 155>,
  nvim_win_get_number = <function 156>,
  nvim_win_get_option = <function 157>,
  nvim_win_get_position = <function 158>,
  nvim_win_get_tabpage = <function 159>,
  nvim_win_get_var = <function 160>,
  nvim_win_get_width = <function 161>,
  nvim_win_hide = <function 162>,
  nvim_win_is_valid = <function 163>,
  nvim_win_set_buf = <function 164>,
  nvim_win_set_config = <function 165>,
  nvim_win_set_cursor = <function 166>,
  nvim_win_set_height = <function 167>,
  nvim_win_set_hl_ns = <function 168>,
  nvim_win_set_option = <function 169>,
  nvim_win_set_var = <function 170>,
  nvim_win_set_width = <function 171>
}

```
</p>
</details>

**Comment by LunarLambda** [comment]
How can I call `nvim_set_client_info` from the Neovim TUI?

I'm currently implementing a sort of proof of concept for a tmux-style start/detach/reattach workflow with Neovim and libuv. I got as far as automatically starting sessions as needed, but I don't know how to uniquely identify the neovim client itself such that it can detach only itself and not everyone else.

The Neovim TUI itself does not seem to call `set_client_info`. Is there any way to maybe map a channel ID to the PID of the client? A simple check like that would be easy to implement.

**Comment by justinmk** [comment]
> but I don't know how to uniquely identify the neovim client itself

Nvim currently uses `stdin_tty || stdout_tty` to guess (with reasonable certainty) that a UI is the TUI: https://github.com/neovim/neovim/blob/7588ff2d8986e343d440dc8e025b1b5d4d8974b5/src/nvim/ui.c#L162-L166

> The Neovim TUI itself does not seem to call `nvim_set_client_info `

Let's change that! Could be done here or related: https://github.com/neovim/neovim/blob/7588ff2d8986e343d440dc8e025b1b5d4d8974b5/src/nvim/ui_client.c#L118-L127

**Comment by LunarLambda** [comment]
I've been thinking about the problem some more and I think the ideal way to do this would be an API function that returns information about the client that made the API request (similar to how many Web APIs have a /me/ or /self/ endpoint that returns information about the API client/authenticated app).

At minimum, this would return the server side of the UI channel that gets returned by `nvim_list_uis`.

The reason `set_client_info` by itself is insufficient is that even if the TUI uniquely identifies itself, the client has to iterate over *all* UIs, then call `get_chan_info` for each one, then look at the `client` info to try and identify itself. But that requires some ability to retrieve/compare the client info against the one retrieved from the API.

However, as far as I can tell, all process related APIs run in the context of the server, not the client. E.g. `uv.os_getpid()` alwats returns the PID of the server, not the client. So I'm not sure where the client would hold onto its unique data to be able to retrieve it during this step.

The server however can very quickly figure out which client sent an API/RPC request, so a 'who am I' function would be fairly straightforward to implement on the server side, and it would sidestep the issue of needing to check the channel of every single UI, and the issue of every client needing to have some way to uniquely identify itself out of the list of UIs, and needing.

Actually, looking at the docs `nvim_get_chan_info` says it takes a 'channel_id, or 0 for current channel'. I believe 0 is currently only applicable in the context of channel callbacks. Maybe `get_chan_info(0)` outside of a callback could be made to return the current API client's channel?

I'm not familiar enough with the exact ins and outs of Neovim's current client/server architecture, but I do think something along the lines of what I described would be the best way to move towards detachable UIs.

**Comment by justinmk** [comment]
> `nvim_get_chan_info` says it takes a 'channel_id, or 0 for current channel'. I believe 0 is currently only applicable in the context of channel callbacks

It allows any API client to get its own client info, without needing to lookup the exact channel id.

**Comment by LunarLambda** [comment]
only the client info? It returned `vim.empty_dict()` for me. If it returned the channel id too, that would be very helpful since it makes calling `chanclose` for detaching trivial.

**Comment by justinmk** [comment]
> It returned `vim.empty_dict()` for me

from what context? you need to call it from a API client. 

> If it returned the channel id too

it does. that's mentioned in the docs.

**Comment by LunarLambda** [comment]
> > It returned `vim.empty_dict()` for me
> 
> from what context? you need to call it from a API client (which implicitly has a channel).
> 
> > If it returned the channel id too
> 
> it does. that's mentioned in the docs.

`:lua =vim.api.nvim_get_chan_info(0)`

I figured this would be equivalent to what I can do from inside of a plugin.

**Comment by zeertzjq** [comment]
You need to use a API client. Calling it inside Nvim will not work.

**Comment by LunarLambda** [comment]
How do I obtain an API client from inside a plugin that runs in the server process? Is there a way to run the plugin code in the TUI client not the server?

**Comment by justinmk** [comment]
please start a thread at https://github.com/neovim/neovim/discussions , this is not relevant to this issue.

**Comment by LunarLambda** [comment]
https://github.com/neovim/neovim/discussions/30314 posted.

**Comment by cosminadrianpopescu** [comment]
> please start a thread at https://github.com/neovim/neovim/discussions , this is not relevant to this issue. the problem that you are trying to solve is not clear.

The issue that @LunarLambda tries to solve is this: try to start a neovim headless (let's call it nvim-server). Then, open in any graphical environment an xterm (let's call it xterm A). from xterm A, connect to nvim-server. 

Now, open another xterm in the same graphical environment (let's call it xterm B). from xterm B, again connect to the same nvim-server. 

Now, @LunarLambda wants to disconnect only from xterm B, leaving xterm A instance connected. 

How can chan_id of xterm B can be associated with the the current running nvim client instance in xterm B, so disconnecting nvim client from xterm B does not disconnect nvim client from xterm A instance.

@LunarLambda correct me if i'm wrong...

**Comment by LunarLambda** [comment]
Yeah that's basically it. See also the discussion I linked. I will move things there so this isn't being talked about in two disparate places

**Comment by justinmk** [comment]
Copying proposal from https://github.com/neovim/neovim/discussions/30314#discussioncomment-10589960 :

> How can a client detach itself?

Specifically you mean the *nvim builtin UI client*. Any other client can call `nvim_get_chan_info(0)` to get its channel id. But the `nvim` builtin UI client doesn't expose a way for users/plugins to execute code from its "local" context.

For the default case where the nvim ui spawns the server as a *child* process, I think it should work if (1) the `nvim` server is spawned with `detach=true`, (2) terminating the *parent* (ui) process should allow the *child* (server) to continue running.

For the non-default case (`nvim --remote-ui`), there are some obvious potential workarounds, such as storing a flag (`nvim --server ./hello.sock --remote-send ':let ui=42<cr>' && nvim --remote-ui ...`).

I see least two clear next steps:

1. `nvim +cmd --remote-ui` should execute `+cmd` in the remote server. No reason for it to behave the way it currently does.
2. For the builtin nvim TUI, "current UI" has no mapping to the list of UIs.
    - If I run `:detach`, how will the server know which UI is the "current UI"?
    - Or a more general solution is to provide a way 
3. the builtin UI should call `nvim_set_client_info` to provide all possible info about itself, including PID.
    - but that still doesn't solve (2).

**Comment by cosminadrianpopescu** [comment]
> For the non-default case (nvim --remote-ui), there are some obvious potential workarounds, such as storing a flag (nvim --server ./hello.sock --remote-send ':let ui=42<cr>' && nvim --remote-ui ...).

I don't think this actually works. Doing `:let ui=42<cr>` sets the flag still on the server side. 

Assuming I am in xterm A, I do there `nvim --server ./hello.sock --remote-send ':let ui=42<cr>' && nvim --remote-ui ...`. Now, in xterm A I am connected to `hello.sock` and the flag is set to 42. 

Now, in xterm B, I do `nvim --server ./hello.sock --remote-send ':let ui=43<cr>' && nvim --remote-ui ...`. Now, in xterm B I am also connected to `hello.sock`. But since the let has been ran on the server, not client side, I am assuming that in both xterm A and xterm B, the value of the flag is 43. So, disconnecting 43 (by doing chanclose on the chan_id identified viat the ui flag == 43), would still disconnect both. 

@justinmk am I wrong? Maybe I misunderstood something?

**Comment by justinmk** [comment]
> Doing `:let ui=42<cr>` sets the flag still on the server side.

The point is that whatever script calls `--remote-send ':let ui=...'` can set a hint that it uses later. It's a hack though.

**Comment by justinmk** [comment]
With https://github.com/neovim/neovim/pull/30319 , `ctrl-z` will detach the current TUI. Please try it out.

**Comment by justinmk** [comment]
#32385 adds a `:detach` command.

**Comment by justinmk** [comment]
#32385 adds a `:detach` command. Followup work is tracked in https://github.com/neovim/neovim/pull/30319 .

I don't know if `:detach` works on Windows, please try it:

> * Define environment variable `__NVIM_DETACH=1` _before_ starting `nvim`
> * run `nvim`
> * does it work normally?
> * try the `:detach` command

**Comment by dundargoc** [comment]
> does it work normally?

On debug it aborts with explicit popup error. On release it just aborts quietly.

**Comment by justinmk** [comment]
Does it always abort, or only if you define `__NVIM_DETACH=1` ?

**Comment by dundargoc** [comment]
Only if `__NVIM_DETACH` is defined.
